src/lib/SaferNames/Name.hs

```diff
@@ -4,267 +4,113 @@
-  where UnsafeMakeName unitRawName = unitName
-
-envMapWithKey :: (Name i -> a -> b) -> Env i a -> Env i b
-envMapWithKey f (UnsafeMakeEnv m) = UnsafeMakeEnv $ LM.mapWithKey f' m
-  where f' rawName = f $ UnsafeMakeName rawName
-
-envPairs :: Env (n:=>:l) a -> (PlainBinderList n l, [a])
-envPairs (UnsafeMakeEnv m) = do
-  let (names, vals) = unzip $ LM.assocs m
-  (unsafeMakePlainBinderList names, vals)
-
-unsafeMakePlainBinderList :: [RawName] -> PlainBinderList n l
-unsafeMakePlainBinderList names = case names of
-  [] -> unsafeCoerceB Empty
-  (name:rest) -> Nest (UnsafeMakeBinder name) $ unsafeMakePlainBinderList rest
-
--- === specialized traversals ===
-
-type NameSet n = S.Set (Name n)
-
-fmapNames :: HasNamesE e => Scope b -> (Name a -> Name b) -> e a -> e b
-fmapNames scope f e = runIdentity $ traverseNamesE scope t e
-  where t = newNameTraversal (pure . f)
-
-foldMapNames :: (HasNamesE e, Monoid a) => (Name n -> a) -> e n -> a
-foldMapNames f e =
-  execWriter $ traverseNamesE unitScope t e
-  where t = newNameTraversal \name -> tell (f name) >> return unitName
-
-freeNames :: HasNamesE e => e n -> NameSet n
-freeNames e = foldMapNames S.singleton e
-
--- should be equivalent to `\ext -> fmapNames (injectName ext)`
-injectNamesL :: HasNamesE e => FreshExt n l -> e n -> e l
-injectNamesL _ = unsafeCoerceE
-
--- TODO: consider returning one of the `Name (n:=>:l)` names found in the
--- `Nothing` case, or even the full set.
-projectNamesL :: HasNamesE e => Scope n -> Scope (n:=>:l) -> e l -> Maybe (e n)
-projectNamesL scope scopeFragment e = traverseNamesE scope t e
-  where t = newNameTraversal \name -> case projectName scopeFragment name of
-                                        Left name' -> Just name'
-                                        Right _    -> Nothing
+(<>>) :: Env v i o -> EnvFrag v i i' o -> Env v i' o
+(<>>) (Env f frag) (EnvFrag frag') = Env f $ concatNameMaps frag frag'
+
+infixl 1 <.>
+(<.>) :: EnvFrag v i1 i2 o -> EnvFrag v i2 i3 o -> EnvFrag v i1 i3 o
+(<.>) (EnvFrag env) (EnvFrag env') = EnvFrag (concatNameMaps env env')
+
+emptyEnvFrag :: EnvFrag v i i o
+emptyEnvFrag = EnvFrag emptyNameMap
+
+lookupEnvFrag :: EnvFrag v i i' o -> Name s i' -> Either (Name s i) (v s o)
+lookupEnvFrag (EnvFrag m) name =
+  case projectName (nameMapNames m) name of
+    Left  name' -> Left name'
+    Right name' -> Right $ lookupNameMap m name'
+
+lookupScope :: Scope n -> Name s n -> s n
+lookupScope (Scope m) v = fromIdE $ lookupNameMap m v
+
+appendScope :: Scope n -> ScopeFrag n l -> Scope l
+appendScope (Scope m1) s2@(ScopeFrag m2) =
+  Scope $ extendNameMap (injectNames s2 m1) m2
 
 -- === common scoping patterns ===
 
 data Abs (binder::B) (body::E) (n::S) where
   Abs :: binder n l -> body l -> Abs binder body n
 
-data FreshAbs (binder::B) (body::E) (n::S) where
-  FreshAbs :: FreshExt n l -> binder n l -> body l -> FreshAbs binder body n
-
 data NestPair (b1::B) (b2::B) (n::S) (l::S) where
```

Comment:
What's an example of a `NestPair`?

Reply:
I'm not sure if we have any in our ASTs. `NestPair b1 b2 n l` is to `Nest b n l` as `(a1, a2)` is to `[a]`, so it seems appropriate to have it for completeness. At various times I've found it useful to create a `NestPair` on the fly to use its instances.

Reply:
Ah, so this is how you say there are two different kinds of binders that are being composed. And I guess `NestPair b1 b2` is itself a `B`?

(I wonder if we could have more descriptive names for `E` and `B`? But I guess it's nice to not have to type it everywhere...)